### PR TITLE
test(ffe): add direct agentless EVP contract

### PR DESCRIPTION
**Pull request by leoromanovsky**
## Motivation

Feature Flags is shipping agentless, CDN-delivered configuration to simplify customer deployments. Exposures and flag-evaluation aggregates also need a reliable network path out of the application: SDKs should prefer an available Agent or serverless proxy and deliver directly to EVP intake when no usable local route exists.

[FFLSDK-183](https://linear.app/datadoghq/issue/FFLSDK-183/evp-fallback-across-all-server-sdks) tracks this across languages and frameworks. This PR provides the shared system-test foundation for direct delivery and a forced `serverless-init` control. It does not claim complete fallback-resilience coverage or enable every SDK.

## Changes

- Reuse exposure and flag-evaluation contracts across Agent, direct, and sidecar topologies; record request, process, network, and shutdown evidence for validation and replay.
- Keep the original default-agentless scenario implicit: both `DD_FEATURE_FLAGS_CONFIGURATION_SOURCE` and `DD_EXPERIMENTAL_FLAGGING_PROVIDER_ENABLED` remain absent. Explicit source/provider settings apply only to the egress scenarios.
- Keep delivery waits in scenario lifecycle handling, not the new flag-evaluation test's post-shutdown assertions. Add regression coverage for missing captures, invalid topology, skipped tests, and empty replay runs.
- Pin `serverless-init` 1.10.4 through the existing content-addressed image catalog.

### Capability matrix

New agentless contracts remain manifest-disabled here; sibling language PRs activate the supported signals and frameworks.

| Capability | Test / assertion | Evidence |
|---|---|---|
| Default agentless configuration | `Test_FFE_Agentless_Configuration` and scenario self-tests | Source and legacy enable flag are absent; evaluation uses configuration from the mock UFC backend. |
| Direct exposures | `Test_FFE_Exposure_Egress_Agentless_Direct` | Five evaluations return the expected variation; one exposure request has the expected flag, variant, allocation, subject, and service context. |
| Direct flag-evaluation counts | `Test_FFE_EVP_Flagevaluation_Egress_Agentless_Direct` | Valid buckets total exactly five evaluations, without duplicate serialized-visible buckets. |
| Forced sidecar control | `Test_FFE_Exposure_Egress_Agentless_Sidecar` and `Test_FFE_EVP_Flagevaluation_Egress_Agentless_Sidecar` | Events arrive on the relay-to-intake capture interface; the direct capture route stays unused. |
| Direct wire contract | `assert_agentless_evp_intake_request` | Intake hostname, port 443, POST, canonical signal path, JSON, API-key header, SDK origin/version, and 202 response; no EVP subdomain header. |
| No-relay topology and route isolation | `assert_direct_evp_runtime_evidence`, `assert_agentless_evp_topology`, `assert_no_evp_proxy_requests` | Direct runs contain only the weblog and TLS observer, with no Agent endpoint variables or Agent/init processes; no sidecar captures or EVP proxy routes. |
| Bounded exposure shutdown | `Test_FFE_Exposure_Egress_Agentless_Direct_Shutdown` | A unique exposure is absent before SIGTERM and captured exactly once after the server-close marker; no explicit `/flush`, clean exit, bounded stop. |
| Replay and disabled-test safety | `test_ffe_evp_contract.py`, `test_mock_ffe_agentless_backend.py` | Persisted evidence replays; manifest-xfailed and empty runs do not acquire impossible capture or artifact obligations. |

### Internal data flow

```mermaid
flowchart LR
    Driver["System-test driver"]
    UFC["Mock UFC backend"]
    subgraph Weblog["Language weblog / SDK"]
        Handler["/ffe handler"] --> Provider["OpenFeature provider"]
        Provider -->|variation| Handler
        Provider --> Exposures["Exposure writer"]
        Provider --> Evaluations["Flag-evaluation writer"]
        Exposures --> Transport["SDK event transport"]
        Evaluations --> Transport
        Shutdown["Runtime shutdown"] --> Provider
    end
    Driver -->|evaluations| Handler
    Driver -->|SIGTERM for exposure shutdown test| Shutdown
    Provider -->|poll| UFC
    UFC -->|configuration| Provider
    Transport -->|direct scenario| Direct["TLS observer: direct capture + mock 202"]
    Transport -->|forced-sidecar scenario| Relay["serverless-init"]
    Relay --> Sidecar["TLS observer: sidecar capture + mock 202"]
    Driver -. "validate captures and runtime evidence" .-> Direct
    Driver -. "validate control route" .-> Sidecar
```

The two event paths are separate test topologies, not a simulated failover sequence. The TLS observer is test infrastructure, not an application delivery relay. Direct runs have no Agent, sidecar, or `serverless-init`; neither path contacts a real Datadog intake.

## Decisions

- Keep **agentless language activation** in sibling PRs. The new ordinary-Agent flag-evaluation contract inherits Go's existing file-level gate.
- Keep configuration, exposures, and flag evaluations as independent acceptance signals. OTLP metrics do not substitute for EVP events.
- Cover steady-state direct delivery and a forced-sidecar control here. Live relay-to-direct failure transitions, v4/v2 negotiation, sticky routing, recovery, redirects, and ambiguous-failure replay remain outside this foundation.
- Identity assertions are **direct-only**; SDK-to-relay credential omission and proxied identity remain SDK-native checks. Shutdown proof is **exposure-only**.
- Use production-shaped HTTPS setup without claiming comprehensive TLS-negative validation: captures assert hostname/port, not TLS negotiation. API-key assertions accept the capture scrubber's redacted value; they do not prove exact secret bytes or backend authorization.

Review focus: scenario-owned lifecycle hooks, replay/disabled-test safety, and whether the direct-topology evidence excludes hidden relays.

## Validation

- Candidate: `af31f51902828b84387c5a7126f6d422e152fe82`; GitHub-verified signature. PR diff base: `f61e30561eaa6b25fd844420edad84cbb7c1f9e7`.
- Local Python 3.12 checks passed on the candidate and on the prior CI merge tree `524bc5c9556b2561acc4a5c15ebcf916980fa8cd` with the lint-fix patch applied: `./format.sh --check` and the focused command below (**74 passed**). Uncached mypy also passed on the patched merge tree.

```sh
python -m pytest -S TEST_THE_TEST \
  tests/test_the_test/test_mock_ffe_agentless_backend.py \
  tests/test_the_test/test_ffe_evp_contract.py \
  tests/test_the_test/test_update_mirror_images.py -q -p no:warnings
```

- Current-head CI tested synthetic merge `f0cb0bf956f8125c6b78230ec8c306a5e631b59c` with main `46453e9d9d1ff266a6188cad4a039038408e8b73`: [full lint passed](https://github.com/DataDog/system-tests/actions/runs/35610327688/job/106367533524); [Test the test passed](https://github.com/DataDog/system-tests/actions/runs/35610327688/job/106367533352) with **596 passed, 1 expected xfail**.
- The broader CI matrix is still running at this update; this is not an all-green or approval claim. No fresh local language E2E builds were run because of workspace disk pressure. Mock-intake evidence is not real-backend acceptance.

### CODE REVIEW

File: utils/_context/_scenarios/agentless_endtoend.py [hunk 1]
```diff
@@ -211,18 +265,394 @@ def _set_components(self) -> None:
...
+        interface.wait(DIRECT_EVP_CAPTURE_SETTLE_SECONDS)
+
+    def _wait_and_stop_containers(self, *, is_empty_test_run: bool) -> None:
+        try:
+            self._wait_for_expected_evp_captures(is_empty_test_run=is_empty_test_run)
+        finally:
+            if self.replay and self.exposure_egress == "direct":
+                self._last_direct_evp_runtime_evidence = json.loads(
+                    self._direct_evp_runtime_evidence_path.read_text(encoding="utf-8")
+                )
```
[the remark this answers is not in the document]
Comment (leoromanovsky) [reply utils/_context/_scenarios/agentless_endtoend.py]: Fixed in dec371d14b85266a33faac1a92ac37b81429a9a6. Direct replay loads runtime evidence only for non-empty runs when the artifact exists. Regression coverage includes both skip-empty and manifest-deactivated selections; the formerly failing replay also passes locally.

File: utils/_context/_scenarios/agentless_endtoend.py [hunk 2]
```diff
@@ -211,18 +265,394 @@ def _set_components(self) -> None:
...
+        )
+        if stop_error is not None:
+            raise stop_error
+
+    def _wait_for_expected_evp_captures(self, *, is_empty_test_run: bool) -> None:
+        """Keep the application alive until selected async EVP writers are observable."""
+        if self.replay or self.exposure_egress is None or is_empty_test_run:
+            return
+        if not self._expected_evp_capture_paths and self._shutdown_evp_evaluation is None:
+            raise RuntimeError("A non-empty Feature Flags EVP run registered no expected capture paths")
```
[the remark this answers is not in the document]
Comment (leoromanovsky) [reply utils/_context/_scenarios/agentless_endtoend.py]: Fixed in dec371d14b85266a33faac1a92ac37b81429a9a6. A run with no registered capture expectations is now a no-op, covering manifest-deactivated items whose setup methods never execute. The regression test exercises this with is_empty_test_run false.

File: utils/_context/_scenarios/agentless_endtoend.py [hunk 3]
```diff
@@ -211,18 +265,394 @@ def _set_components(self) -> None:
...
+            events = content.get("flagEvaluations")
+        if not isinstance(events, list):
+            return False
+        return any(
+            isinstance(event, dict)
+            and isinstance(event.get("flag"), dict)
+            and event["flag"].get("key") == evaluation.flag_key
+            and isinstance(event.get("subject"), dict)
+            and event["subject"].get("id") == evaluation.subject_id
+            for event in events
```
[the remark this answers is not in the document]
Comment (leoromanovsky) [reply utils/_context/_scenarios/agentless_endtoend.py]: Fixed in dec371d14b85266a33faac1a92ac37b81429a9a6. The matcher now branches by signal path and accepts both raw and sha256-prefixed flag-evaluation targeting keys. Added parameterized regression coverage for both forms.

File: utils/_context/_scenarios/agentless_endtoend.py [hunk 4]
```diff
@@ -129,7 +172,16 @@ def __init__(
...
+        self._expected_evp_capture_paths: set[str] = set()
+        self._shutdown_evp_evaluation: DirectEVPShutdownEvaluation | None = None
+        self._last_direct_evp_runtime_evidence: dict[str, Any] | None = None
+        self._last_direct_evp_shutdown_evidence: dict[str, Any] | None = None
         environment: dict[str, str | None] = {
+            # The shared weblogs use this switch to install their OpenFeature provider. The
+            # configuration source selects how the provider receives flags; it does not make the
+            # application adopt the provider on its own.
+            "DD_EXPERIMENTAL_FLAGGING_PROVIDER_ENABLED": "true",
+            "DD_FEATURE_FLAGS_CONFIGURATION_SOURCE": "agentless",
```
[the remark this answers is not in the document]
Comment (leoromanovsky) [reply utils/_context/_scenarios/agentless_endtoend.py]: Addressed in [793233f0f7](https://github.com/DataDog/system-tests/commit/793233f0f70bf49a37c0a3d2c8804fde1937a7df). The original Agentless configuration scenario now leaves DD_FEATURE_FLAGS_CONFIGURATION_SOURCE unset; only the direct/sidecar egress scenarios set it explicitly. Regression coverage requires the source to be absent in the original scenario and explicit in both egress scenarios. Resolving as fixed in this PR.

File: tests/ffe/utils/evp.py [hunk 5]
```diff
@@ -0,0 +1,447 @@
...
+    )
+
+    weblog = containers_by_name["system-tests-weblog"]
+    assert weblog.get("library") == library_name, (
+        f"Live weblog library label is {weblog.get('library')!r}, expected {library_name!r}"
+    )
+    expected_executables = EVP_LANGUAGE_PID1_EXECUTABLES.get(library_name)
+    assert expected_executables, f"No ordinary PID 1 contract is defined for {library_name!r}"
+    pid1_executable = Path(str(weblog["pid1_command"]).split()[0]).name
+    assert _is_expected_pid1_executable(library_name, pid1_executable, expected_executables), (
```
[the remark this answers is not in the document]
Comment (leoromanovsky) [reply tests/ffe/utils/evp.py:289]: Addressed in the stacked [.NET enablement PR #7707](https://github.com/DataDog/system-tests/pull/7707), at [ccf370de4c](https://github.com/DataDog/system-tests/commit/ccf370de4c3e4dc62aaf09977d2727d063449b05). Its opt-in direct-test launcher uses exec dotnet so the application is PID 1, with a regression test guarding launcher ordering. We retain the strict shared topology assertion rather than weakening it. This language-specific fix is implemented in the child PR (still open), not merged into this base; resolving here with that dependency explicit.

File: utils/_context/_scenarios/agentless_endtoend.py [hunk 6]
```diff
@@ -169,9 +221,14 @@ def __init__(
         )
 
         if exposure_egress == "direct":
+            self.weblog_infra.library_container.environment[DIRECT_EVP_SHUTDOWN_FLUSH_ENV] = "true"
+            self.weblog_infra.library_container.volumes[DIRECT_EVP_CA_BUNDLE_SOURCE] = {
+                "bind": DIRECT_EVP_CA_BUNDLE_CONTAINER_PATH,
+                "mode": "ro",
```
[the remark this answers is not in the document]
Comment (leoromanovsky) [reply utils/_context/_scenarios/agentless_endtoend.py:227]: Addressed in the stacked [Java enablement PR #7702](https://github.com/DataDog/system-tests/pull/7702), at [364b7dcfd7](https://github.com/DataDog/system-tests/commit/364b7dcfd7294b0ede91222b688ca90d8e9d52dd). The direct spring-boot launcher copies the Java truststore, imports the observer CA, and sets the JSSE truststore options while preserving JAVA_OPTS and normal startup. Regression tests cover that wiring. This language-specific fix is implemented in the child PR (still open), not merged into this base; resolving here with that dependency explicit.

**Review by vjfridge — APPROVED** [review]
Code looks solid, test coverage looks good, and remaining CI failures seem unrelated to the changes in this PR.

LGTM!

File: tests/ffe/test_flag_eval_evp.py [hunk 7]
```diff
@@ -237,6 +255,103 @@ def assert_no_duplicate_visible_events(events: list[tuple[JSON, JSON]]) -> None:
...
+        ]
+        assert matching_requests, f"Expected flagevaluation requests for {self.flag_key}"
+
+        events = [
+            event
+            for request in matching_requests
+            for event in evp_flagevaluation_events_from_data(request, self.flag_key, self.targeting_key)
+        ]
+        assert events, f"Expected EVP flagevaluation events for flag {self.flag_key}"
+        for _, event in events:
```
Comment (typotter) [inline tests/ffe/test_flag_eval_evp.py:306]: Is it worth checking the context here too? I can't contrive a case where the incoming events would have the right count but malformed/unxpected context though.

**Review by typotter — APPROVED** [review]

File: tests/test_the_test/test_ffe_evp_contract.py
Comment (nccatoni) [inline tests/test_the_test/test_ffe_evp_contract.py:1]: The `test_the_test` directory is intended to test the system-tests framework. I think testing the test code itself may be unnecessary, but if you want to do it, the best approach would be to put it next to the code you’re testing, perhaps in `tests/ffe/<testing_tests|something_else>`. I’m not aware of any test like that currently in the repo, so there is no canonical way to do it. The most important thing is not to use the `TEST_THE_TEST` scenario because it runs on every PR in system-tests. It’s important to keep it focused on framework assertions.

File: tests/test_the_test/scenarios.json
Comment (nccatoni) [inline tests/test_the_test/scenarios.json:1]: Just FYI but you can leave it, you don't need to update this file every time you add a test it's just mocked data for a framework test it doesn't need to be in sync with the actual content of tests/.

**Review by nccatoni — CHANGES_REQUESTED** [review]
Outside of the comment about test_the_test the rest looks good to me

**Review by nccatoni — APPROVED** [review]
